### ISSUE TITLE
Always unlocks with no checks

### ISSUE THREAD

**Issue by alba4k** [ChocolateLoverRaj/pam-any#12]
I've put the following in `/etc/pam.d/sudo` on arch

```

auth sufficient libpam_any.so { "mode": "One", "modules": { "login": "Password", "pam_howdy": "Face Recognition", "pam_fprintd": "Fingerprint" } }

auth		include		system-auth
account		include		system-auth
password	include		system-auth
```

And it looks like it's always just unlocking without checking either of those three options, immediately. Any idea of what could be causing this?

**Comment by alba4k** [comment]
Looked into pam-random

What I put there as modules were the names of the files in /lib/security, removing .so from them. Do i need to create separate files in /etc/pam.d instead, that only have a single `auth sufficient` followed by the .so file I want to use with them?

**Comment by alba4k** [comment]
Nope, that does not seem to be the issue. Only having the following also causes sudo to run withotu checking anything

![immagine](https://github.com/user-attachments/assets/2d26f40c-8542-4d5b-a4a7-80e29dae846d)

**Comment by ChocolateLoverRaj** [comment]
Does this happen with `pamtester sudo $USER authenticate`?

**Comment by alba4k** [comment]
Yep it does

`pamtester: succesfully authenticated`

**Comment by alba4k** [comment]
@ChocolateLoverRaj ever got a chance to look into this?

**Comment by ChocolateLoverRaj** [comment]
Sorry, I forgot about this. I'm currently working on something else. I don't personally use this. I might look into this later. If you or anyone else wants to investigate this feel free to do so since there is a very small amount of code.

**Comment by outofrange** [comment]
Having the same issue on NixOS, both with the example config, as well as one without any modules at all: `auth required /nix/store/61napg11y0rxq2gkwbwcpngpbb9wcswn-pam-any-0.1.0/lib/libpam_any.so {"mode":"One","modules":{}}`

Looking at the code: `PAM_SUCCESS` requires at least 1 module to `result.is_ok()`.
But how is this possible when 0 modules are passed - who is answering?

Edit:
Last line executed appears to be [`let user = pam_try!(pamh.get_user(None));`](https://github.com/ChocolateLoverRaj/pam-any/blob/e77687709d092a6bb77e57a444403798855075c0/src/lib.rs#L50) - logs placed after this line aren't written to stdout anymore.

**Comment by ChocolateLoverRaj** [comment]
> Having the same issue

@outofrange  Does it instantly unlock with `{"mode":"One","modules":{}}`? Because it should instantly fail. Also using `{"mode":"One","modules":{}}` doesn't make sense because you aren't specifying any PAM modules. I think failing by default makes more sense than authenticating by default.

**Comment by outofrange** [comment]
I tried it with `auth sufficient libpam_any.so { "mode": "One", "modules": { "login": "Password", "pam-random": "Random Chance" } }`, and `auth sufficient libpam_any.so { "mode": "One", "modules": { "login": "Password" } }`, both with and without `account sufficient libpam_any.so`; 
`{"mode":"One","modules":{}}` was just a test to see if modules are called at all, expecting it to fail - which it didn't.

> I think failing by default makes more sense than authenticating by default.

100% agree!

But, as mentioned, it seems to get to `pam_try!(pamh.get_user(None))`, which appears to be failing - but it's still being treated as authenticated.
That also explains why it doesn't matter which modules I configure, or if I configure any at all - the code doesn't even get that far.

I need to investigate why `pam_try!(pamh.get_user(None))` doesn't work for me, could be some misconfiguration on my end with NixOS and all.
But even then, I don't think that errors with getting the user should result in an authentication success - so I'm actually seeing two issues here:
- modules aren't called, probably because some issue with calling `pamh.get_user(None)` (this issue here)
- erros while calling `pamh.get_user(None)` are treated as successful authentication (should I create a separate issue for this?)

**Comment by ChocolateLoverRaj** [comment]
Btw the PAM stuff relies on https://github.com/anowell/pam-rs. It seems to be not getting updates anymore so that might cause the issue.

What is the return value of `pamh.get_user(None)`?

What happens if you do `pam_try!(pamh.get_user(None), PAM_AUTH_ERR)`?

**Comment by FakeShell** [comment]
i was facing the same issue. so i ended up rewriting pam any:
https://github.com/furiLabs/pam-parallel

you can find a demonstration of it using our biomd fingerprint service here:
https://social.likeadragonmaid.dev/@fakeshell/114277472722446338

if anyone is interested then pam-parallel is available too. tested in debian trixie

**Comment by ChocolateLoverRaj** [comment]
@FakeShell if more people try ur `pam-parallel` and they like it then I will add a link in the README to recommend people to use urs instead.

Also your demo looks very cool.

**Comment by alba4k** [comment]
@FakeShell mind sharing some instructions on how this should be used?

Compiling it and installing to `/lib/security/pam_parallel.so` was no issue, obviously, but what should I put in `/etc/pam.d/sudo`?
`auth sufficient pam_parallel.so { "mode": "One", "modules": { "login": "Password", "pam_fprintd.so": "Fingerprint" } }` isn't working

**Comment by FakeShell** [comment]
> [@FakeShell](https://github.com/FakeShell) mind sharing some instructions on how this should be used?
> 
> Compiling it and installing to `/lib/security/pam_parallel.so` was no issue, obviously, but what should I put in `/etc/pam.d/sudo`? `auth sufficient pam_parallel.so { "mode": "One", "modules": { "login": "Password", "pam_fprintd.so": "Fingerprint" } }` isn't working

in /etc/pam.d/biomd i have:
```
auth    requisite       pam_biomd.so debug
account required        pam_permit.so
```

and /etc/pam.d/login is untouched debians default configuration (likely the same on other distros too)
then in sudo i have:
```
#%PAM-1.0

auth    sufficient pam_parallel.so debug { "mode": "One", "modules": {"biomd": "Fingerprint", "login": "Password"} }

@include common-auth
@include common-account
@include common-session-noninteractive
```

obviously you can swap out biomd for other methods. you need to pass it a pam service, not a shared object. so if your Fingerprint is fprintd, then you need a pam_fprintd in /etc/pam.d/ and you can pass it like this:
```
auth    sufficient pam_parallel.so debug { "mode": "One", "modules": {"pam_fprintd": "Fingerprint", "login": "Password"} }
```

the argument you pass is supposed to be a service in /etc/pam.d/ not a module in /usr/lib/$(TRIPLET)/security/

the debug argument logs stuff so you can follow the process with journalctl -f which will tell you about possible issues such as syntax errors

i also added a test configuration here
https://github.com/FuriLabs/pam-parallel/blob/trixie/data/pam_parallel
which you can test using pamtester:
```
pamtester pam_parallel $USER authenticate
```

**Comment by jfly** [comment]
This is an issue in the underlying [pam-bindings crate](https://crates.io/crates/pam-bindings) (which confusingly lives in the [anowell/pam-rs github repo](https://github.com/anowell/pam-rs)). The issue apparently only arises when building in release mode: https://github.com/anowell/pam-rs/issues/16.

@ChocolateLoverRaj, as you mentioned above, pam-rs seems to be unmaintained. [nonstick](https://crates.io/crates/nonstick) appears to be a maintained fork of it. Would you accept a PR porting `pam-any` to use `nonstick`?

**Comment by ChocolateLoverRaj** [comment]
> [@ChocolateLoverRaj](https://github.com/ChocolateLoverRaj), as you mentioned above, pam-rs seems to be unmaintained. [nonstick](https://crates.io/crates/nonstick) appears to be a maintained fork of it. Would you accept a PR porting `pam-any` to use `nonstick`?

Yes ❤️

**Comment by alba4k** [comment]
btw I don't really need this to be fixed anymore since hyprlock now natively supports running fprintd in parallel with pam, so I will unsubscribe

but ping me when closing as I might come back in the future, I honestly believe there is a usecase for something like this

**Comment by lvkv** [comment]
Hi! I've taken over ownership of `pam-bindings` and the root cause of this issue (i.e., https://github.com/lvkv/pam-rs/issues/16) is now fixed in version 0.1.2, which is now on crates. Feel free to open more issues if you run into any problems!

**Comment by ChocolateLoverRaj** [comment]
I created #14 which updates `pam-bindings`. Once someone tries it out and it works we can merge it and close this.
